### DISCUSSION TITLE
Default behavior of Renovate from "eager" to "lazy"?

### DISCUSSION THREAD

**Discussion by ringods** [renovatebot/renovate#29779] [Request Help]
### What would you like help with?

I would like help with my configuration

### How are you running Renovate?

Mend Renovate hosted app on github.com

### Please tell us more about your question or problem

When I enable Renovate in a Github repository, the app is very eager to get to work and create quite a lot of PRs. When I don't want upgrade PRs for certain dependencies, I need to either `exclude...` or `enabled: false` on matching files.

Is there a way to change the default behaviour of Renovate to do nothing and only check what is in my config file? If so, how to I define that in a `renovate.json` file?

**Comment by rarkins** [comment]
Sounds like setting dependencyDashboardApproval=true could be a good fit here. Then you can open PRs on demand using the issue

**Reply by ringods** [reply]
Thank you for the suggestion, but I would rather not have to manually approve the creation of PRs.

**Reply by rarkins** [reply]
You'd prefer to add them to your config one by one? Not sure I understand your use case

**Reply by HonkingGoose** [reply]
# Summary

- Try Renovate's default behavior for a month
- Read the Noise Reduction, and Upgrade Best practices docs before you start writing custom config
- Only write custom config when you _really_ need it
- If you really want to ignore dependencies, use [`dependencyDashboardApproval`](https://docs.renovatebot.com/configuration-options/#dependencydashboardapproval) instead of `exclude` or `enabled:false`

## Try Renovate's default config for a month first

> You'd prefer to add them to your config one by one? Not sure I understand your use case

Building your Renovate configuration one by one sounds painful to me. And you may not need all those custom rules once you're up-to-date and updating regularly.

I recommend you try the default behavior of Renovate for at least a month or so, _before_ you start working on your own custom configuration.

Read our [Renovate docs, Upgrade best practices](https://docs.renovatebot.com/upgrade-best-practices/) for more on our recommendations for updating. In short: if it hurts, do it more often.

## Only write custom rules when really needed

> Thank you for the suggestion, but I would rather not have to manually approve the creation of PRs.

Maybe you meant something like: "Renovate should follow these rules for package `x`"? If yes, you can use `packageRules` for that. Again: I still recommend you try the default behavior first. Writing custom rules is easy to mess up, and may lead you to sub-optimal update practices.

Examples of the kinds of things you can configure with `packageRules`, combined with other config options:

- For package `x` only give me updates once a week, _and_ only after the package is at least `3` days old.
- For package `y` I want to get updates manually from the Dependency Dashboard.
- For package `z` I only want `major` type updates.
- For development type dependencies like Prettier, ESLint, Standard, TypeScript, give me a update once a month

## Prefer Dashboard approval over exclude or enabled:false

> When I enable Renovate in a GitHub repository, the app is very eager to get to work and create quite a lot of PRs. When I don't want upgrade PRs for certain dependencies, I need to either `exclude...` or `enabled: false` on matching files.

I agree with `@rarkins`, to "ignore" dependencies, please use `dependencyDashboardApproval=true`. Or set a "slower `schedule`". If you exclude or disable updates for a dependency Renovate will be _useless_ for that package!

When you use the Dependency Dashboard Approval workflow you can always get PRs on-demand from the Dashboard for the "ignored dependency". I recommend you check the dashboard regularly if you're going to use the Dependency Dashboard Approval workflow.

## Renovate is updating too much!

> Is there a way to change the default behaviour of Renovate to do nothing and only check what is in my config file? If so, how to I define that in a `renovate.json` file?

I don't understand what you want Renovate to do here. You're installing a dependency update bot, so a reasonable default behavior is for Renovate to give you updates.

Logically if you install Renovate into a repository with lots of old dependencies, Renovate will also create about the same number of PRs.

Renovate groups packages that need to be updated together into a single PR, provided the maintainers/community told Renovate about the grouping. So it's not quite as bad as 1 PR for each dependency, usually.

If after a month with the default behavior, you still find Renovate too noisy, read the [Renovate docs, Noise Reduction](https://docs.renovatebot.com/noise-reduction/) to learn about:

- Package Grouping
- Scheduling when Renovate creates PRs
- Automerging PRs when your tests pass
- Lockfile considerations

**Reply by ringods** [reply]
@rarkins @HonkingGoose my use case is to only upgrade a subset of the dependencies which are required in a Pulumi provider repository.

The general layout of a Pulumi provider is like this:

```
.
|- examples
|  |- go.mod
|  |- dotnet/*
|  |- go/*
|  |- python/*
|  \- ts/*
|- provider
|  |- go.mod
|  \- <sources>
\- sdk
   |- dotnet/*
   |- go/*
   |- nodejs/*
   \- python/*
```

* `examples` contains small projects from a user perspectives. Such projects (can) contain a dependency to the project itself. If a newer version of the package is released, these examples may be upgraded.
* `provider` is a bit different. Pulumi has a tool named [upgrade-provider](https://github.com/pulumi/upgrade-provider/). This tool can upgrade some Pulumi specific dependencies to a consistent set:
  * `github.com/pulumi/pulumi-terraform-bridge/v3`
  * `github.com/pulumi/pulumi/sdk/v3`
  * `github.com/pulumi/pulumi/pkg/v3`
  * dependency pointing to upstream TF provider
 The first one also depends on a specific version of the second and third, so `upgrade-provider` does a better job here. It also includes some code migrations. Pulumi has a way to wrap a Terraform provider and expose it as a Pulumi provider. `upgrade-provider` also helps out here to upgrade it with required Pulumi semantics, not just Go semantics. These dependencies should not be handled by Renovate.
* `sdk` contains generated language specific SDKs. The code generation tools from Pulumi take care of setting the *minimum* version of a dependency required for the code to work with users. Renovate shouldn't do anything in here.

Given the above, I have a feeling that I have to disable, exclude, ... more than what is enabled. Therefor, I was searching for a way to switch Renovate's behavior the other way around and only configure what I needed.

## Example

Repository: https://github.com/pulumiverse/pulumi-scaleway/

With the current rule in `packageRules` of `.github/renovate.json`

```json
        {
            "matchFileNames": [
                "provider/**"
            ],
            "matchDatasources": [
                "go"
            ],
            "excludePackagePrefixes": [
                "github.com/hashicorp/terraform-plugin-sdk",
                "github.com/pulumi/pulumi",
                "github.com/scaleway/terraform-provider-scaleway"
            ]
        },
```

 Renovate still generated PR [#308](https://github.com/pulumiverse/pulumi-scaleway/pull/308). Specific to the `go` datasource, how should I write down any exclusions?

**Reply by HonkingGoose** [reply]
Maybe this feature would help?

- #29809

**Reply by rarkins** [reply]
@ringods Renovate updates all packages by default, as you seem to know. To switch that, you add a rule like this:

```
{
  "matchPackagePatterns": ["*"],
  "enabled": false
}
```

Then you would selectively re-enable using a rule like you proposed:

```
        {
            "matchFileNames": [
                "provider/**"
            ],
            "matchDatasources": [
                "go"
            ],
            "excludePackagePrefixes": [
                "github.com/hashicorp/terraform-plugin-sdk",
                "github.com/pulumi/pulumi",
                "github.com/scaleway/terraform-provider-scaleway"
            ],
            "enabled": true
        },
```

The rule as you tried was "pointless" because it did nothing. Your rule said "match X and Y, exclude Z...." but didn't then say _what to do_. excludeX just means "and don't match". It doesn't mean "ignore these packages".

**Reply by ringods** [reply]
@rarkins using the combination of `exclude...` and `enabled: false` reads to me like a double negation. I would interpret your rule as: disable all Go updates in `providers/**` unless they match the 3 listed packages.

**Reply by rarkins** [reply]
You're right, it was a typo - corrected it from false to true
